### PR TITLE
whisperspeak

### PR DESCRIPTION
**Pull request by MacMeDan**
My ObjectiveC is super rusty so let me know if there is anything you would like me to fix.

Add Support for WhisperSpeak.

### CODE REVIEW

**Comment by MacMeDan** [comment]
Resolves Issue #6

File: TSKit.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -668,7 +670,7 @@
 				HEADER_SEARCH_PATHS = "$(PROJECT_DIR)/TSKit/include";
 				INFOPLIST_FILE = TSKit/Info.plist;
 				INSTALL_PATH = "$(LOCAL_LIBRARY_DIR)/Frameworks";
-				IPHONEOS_DEPLOYMENT_TARGET = 10.3;
+				IPHONEOS_DEPLOYMENT_TARGET = 12.2;
```
Comment (JanC) [inline TSKit.xcodeproj/project.pbxproj]: Is there a particular reason you change the deployment target to iOS 12?

File: TSKit.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -678,6 +680,7 @@
 				PRODUCT_NAME = "$(TARGET_NAME:c99extidentifier)";
 				SKIP_INSTALL = YES;
 				TARGETED_DEVICE_FAMILY = "1,2";
+				VALID_ARCHS = "arm64 arm64e armv7 armv7s x86_64";
```
Comment (JanC) [inline TSKit.xcodeproj/project.pbxproj:683]: Same question for those settings

File: TSKit/Public/TSClient.h [hunk 3]
```diff
@@ -83,6 +83,29 @@ typedef void (^TSClientAuthPrompt)(TSClientAuthCallback authCallback);
 
 -(void)disconnect;
 
+/**
+ A client with a whisper list set can talk to the specified clients and channels bypassing the standard rule that voice is only transmitted to the current channel. Whisper lists can be defined for individual clients. A whisper list consists of an array of client IDs and/or an array of channel IDs.
+
+ @param users list of `TSUsers` to be white listed. To clear list pass `Null` or empty array.
+ @param channels List of `TSChannels` to be white listed. To clear list pass `Null` or empty array.
+ */
+-(void)wisperConnect:(nullable NSArray<TSUser *>*)users channels:(nullable NSArray<TSChannel *>*)channels;
```
Comment (JanC) [inline TSKit/Public/TSClient.h]: There is a typo in those new methods (missing `h`)

File: TSKit/Public/TSClient.m [hunk 4]
```diff
@@ -149,6 +149,47 @@ - (void)disconnect
     self.isSubscribedToChannels = NO;
 }
 
+- (void)wisperConnect:(nullable NSArray<TSUser *>*)users channels:(nullable NSArray<TSChannel *>*)channels {
+    UInt64 *channelIDs =calloc(users.count + 1, sizeof(UInt64));
+    for (int i = 0; i < users.count; i++) {
```
Comment (JanC) [inline TSKit/Public/TSClient.m]: you should be iterating over the `channels` and not `users`

Comment (JanC) [inline TSKit/Public/TSClient.m]: `NSUInteger i = 0;` instead of `int i = 0;`

File: TSKit/Public/TSClient.m [hunk 5]
```diff
@@ -149,6 +149,47 @@ - (void)disconnect
...
+            NSLog(@"WisperConnect Success");
+            return;
+        } else {
+            // error
+            NSLog(@"WisperConnect Error: %@", [NSError ts_errorMessageFromCode:errorCode]);
+        }
+    };
+
+    ts3client_requestClientSetWhisperList(self.serverConnectionHandlerID, self.ownClientID, channelIDs, idBuffer, [returnCode cStringUsingEncoding:NSUTF8StringEncoding]);
+
```
Comment (JanC) [inline TSKit/Public/TSClient.m]: There is a memory leak as both `channelIDs` and `idBuffer` are `calloc`ated but not freed. You need to
```c
    free(channelIDs);
    free(idBuffer);
```

File: TSKit/Public/TSClient.m [hunk 6]
```diff
@@ -149,6 +149,47 @@ - (void)disconnect
     self.isSubscribedToChannels = NO;
 }
 
+- (void)wisperConnect:(nullable NSArray<TSUser *>*)users channels:(nullable NSArray<TSChannel *>*)channels {
```
Comment (JanC) [inline TSKit/Public/TSClient.m]: The method name is misleading as it does not "connect".  Maybe you can use the underlaying C function name:
```
-(void) requeSetWhisperList(....`
```

**Review by JanC — CHANGES_REQUESTED** [review]

**Comment by JanC** [comment]
Thanks for the PR!  I left some comments

File: TSKit.xcodeproj/project.pbxproj (hunk 1 above)
Comment (MacMeDan) [reply TSKit.xcodeproj/project.pbxproj]: No specific reason just habit. I can switch it back but I would also ask any specific reason to leave it a 10.3? Just out of curiosity.

File: TSKit/Public/TSClient.m (hunk 5 above)
Comment (MacMeDan) [reply TSKit/Public/TSClient.m]: Thanks for this. I was definitely out of my comfort zone on this.

File: TSKit/Public/TSClient.m (hunk 4 above)
Comment (MacMeDan) [reply TSKit/Public/TSClient.m]: Ahh nice catch.

File: TSKit.xcodeproj/project.pbxproj (hunk 1 above)
Comment (JanC) [reply TSKit.xcodeproj/project.pbxproj]: The projects using the library would be forced to target at least the same iOS version. Even if the iOS 12 adoption is quite high, I'd keep the library deployment low

**Pull request closed without merging**